### ISSUE TITLE
[BUG] @Published field is not propagating to view in iOS 14.5

### ISSUE THREAD

**Issue by skywalkerdude** [fermoya/SwiftUIPager#218]
**Describe the bug**
@Published field is not propagating to view in iOS 14.5

**To Reproduce**
Reproduce with three simple classes:

```
struct PagerView: View {

    @StateObject var page: Page = .withIndex(5)

    var body: some View {
        Pager(page: page,
              data: Array(0..<20),
              id: \.self,
              content: { index in
                DummyView(viewModel: DummyViewModel(num: index))
              })
    }
}
```

```
struct DummyView: View {

    @ObservedObject private var viewModel: DummyViewModel

    init(viewModel: DummyViewModel) {
        self.viewModel = viewModel
    }

    var body: some View {
        VStack {
            Text("Number \(viewModel.num)")
            if viewModel.isLoaded {
                Text("Is Loaded")
            } else {
                Text("Not loaded")
            }
        }.onAppear {
            viewModel.load()
        }
    }
}
```

```
class DummyViewModel: ObservableObject {

    @Published var isLoaded: Bool = false

    let num: Int

    init(num: Int) {
        self.num = num
    }

    func load() {
        isLoaded = true
    }
}
```

**Expected behavior**
`onAppear` calls `viewModel.load()` which sets the `@Published` `isLoaded` field to true. This should cause every page to display `Is Loaded`. However, only the very first page is `Is Loaded`, while everything is is `Not Loaded`. This is because for the pages that are not in the foreground, the published change is not being pushed to the view.

**Screenshots / Videos**

**Environment:**
 - iOS 14.5
 - iPhone 11 pro simulator
- SwiftUIPager version 2.2.0

https://user-images.githubusercontent.com/1427524/117780514-8d8e9380-b1f4-11eb-9483-99177a9fa61e.mov

**Additional context**

**Comment by skywalkerdude** [comment]
Note: If you put a log statement or breakpoint, you'll see `onAppear` **is** being called on the background pages as well

**Comment by fermoya** [comment]
@skywalkerdude the reason it works with a `StateObject` and not `ObservedObject` is the SwiftUI's State management 😅. Take a look at [this answer](https://stackoverflow.com/questions/62544115/what-is-the-difference-between-observedobject-and-stateobject-in-swiftui) in SO.

Your options are:
- Use `StateObject` in your `DummyView` 
- Or ensure its owner/parent holds a `StateObject`

Whenever you scroll `Pager`, the pages gets recreated and you get a new instance of your ViewModel which by default has `isLoaded` set to `false`. This is a common issue in SwiftUI

**Comment by skywalkerdude** [comment]
Hi, sorry. 

I looked into this extensively and I don't think this has to do with SwiftUi state management, but it looks to me it has to do with how `Pager` is rendering the pages:

I added some log statements:

```
class DummyViewModel: ObservableObject {

    @Published var isLoaded: Bool = false

    let num: Int

    init(num: Int) {
        self.num = num
    }

    func load() {
        print("dummy_view_model_log: loading \(num)")
        isLoaded = true
    }
}
```

```struct DummyView: View {

    @ObservedObject private var viewModel: DummyViewModel

    init(viewModel: DummyViewModel) {
        self.viewModel = viewModel
    }

    var body: some View {
        print("dummy_view_log: rendering view \(viewModel.num), isLoaded: \(viewModel.isLoaded)")
        return VStack {
            Text("Number \(viewModel.num)")
            if viewModel.isLoaded {
                Text("Is Loaded")
            } else {
                Text("Not loaded")
            }
        }.onAppear {
            viewModel.load()
        }
    }
}
```

When I run the app and load he first page, here are the logs:
```
dummy_view_log: rendering view 3, isLoaded: false
dummy_view_log: rendering view 4, isLoaded: false
dummy_view_log: rendering view 5, isLoaded: false
dummy_view_log: rendering view 6, isLoaded: false
dummy_view_log: rendering view 7, isLoaded: false
dummy_view_model_log: loading 7
dummy_view_model_log: loading 6
dummy_view_model_log: loading 5
dummy_view_model_log: loading 4
dummy_view_model_log: loading 3
dummy_view_log: rendering view 3, isLoaded: true
dummy_view_log: rendering view 4, isLoaded: true
dummy_view_log: rendering view 5, isLoaded: true
dummy_view_log: rendering view 6, isLoaded: true
dummy_view_log: rendering view 7, isLoaded: true
```

When I make one swipe to page 16, here are the new logs:

```
dummy_view_log: rendering view 3, isLoaded: false
dummy_view_log: rendering view 4, isLoaded: false
dummy_view_log: rendering view 5, isLoaded: false
dummy_view_log: rendering view 6, isLoaded: false
dummy_view_log: rendering view 7, isLoaded: false
dummy_view_log: rendering view 3, isLoaded: false
dummy_view_log: rendering view 4, isLoaded: false
dummy_view_log: rendering view 5, isLoaded: false
dummy_view_log: rendering view 6, isLoaded: false
dummy_view_log: rendering view 7, isLoaded: false
dummy_view_log: rendering view 3, isLoaded: false
dummy_view_log: rendering view 4, isLoaded: false
dummy_view_log: rendering view 5, isLoaded: false
dummy_view_log: rendering view 6, isLoaded: false
dummy_view_log: rendering view 7, isLoaded: false
dummy_view_log: rendering view 3, isLoaded: false
dummy_view_log: rendering view 4, isLoaded: false
dummy_view_log: rendering view 5, isLoaded: false
dummy_view_log: rendering view 6, isLoaded: false
dummy_view_log: rendering view 7, isLoaded: false
dummy_view_log: rendering view 4, isLoaded: false
dummy_view_log: rendering view 5, isLoaded: false
dummy_view_log: rendering view 6, isLoaded: false
dummy_view_log: rendering view 7, isLoaded: false
dummy_view_log: rendering view 8, isLoaded: false
dummy_view_model_log: loading 8
dummy_view_log: rendering view 8, isLoaded: true
```
It looks like pager re-creates each view 3 times but doesn't call `onAppear` on any view except view 8.

I took a look at your SO link, and while I understand that the @ObservedObject is being recreated, it shouldn't necessarily be a problem is `onAppear` is called each time the drawing is complete, so the `isLoaded` boolean is set to true. Is it expected that the `onAppear` call doesn't get triggered after the very first time the page is rendered?

**Comment by fermoya** [comment]
Hi @skywalkerdude , I think you don't understand how SwiftUI works. Your view "appears" off/-screen so its `onAppear` gets called and its reference to your ViewModel gets "loaded". When `Pager` swipes the pages, the state of this view changes and its subviews get recreated. The way this works in `SwiftUI` is the the view's body gets recreated and the `init` gets called and recomputed. But it's `onAppear` doesn't get called again!!! `Pager` or any other view in SwiftUI won't invoke `onAppear` that's reserved to the system.

As mentioned in my previous comments, your problem is you're defining your ViewModel as a `ObservedObject`. If you've read the post I sent you you'll understand why this isn't the right choice for you. Proof of that is switching to `StateObject` worked.
Alternatively, you could hold an array of ViewModels in the view that defines `Pager`.

I'd suggest you use a simple example where you drag a custom view and observe when `onAppear` is called and when how many times `init` is invoked.

**Comment by fermoya** [comment]
@skywalkerdude not sure what you're trying to achieve here. Do you wanna know what page indexes are visible in the screen?

**Comment by skywalkerdude** [comment]
You're right that it's possible I don't fully understand SwiftUI, since I mostly hail from the Android world. I read what you said many times, and it makes sense to me. In fact, that was my initial understanding as well, until I started second guessing myself because of this bug in iOS 14.5.

To answer your second question, I have an app with a rather complicated set of views that uses your pager library (https://github.com/HymnalDreamTeam/Hymns-iOS), and it suddenly stopped working with iOS 14.5, while it worked fine before.  The data stopped loading and it just has an indefinite loading screen (worked fine before). I haven't heard of Apple changing their SwiftUI State management from 14.4 to 14.5, so it's certainly bizarre that it suddenly stopped working.

You're under no obligation to help me debug my app :), but if you are interested, I can certainly give more detail. For now, I may just disable the paging feature temporarily as I try to work through and figure out exactly what is wrong.

**Comment by fermoya** [comment]
Thanks for sharing, I'll give it a look.

Ive experienced changes sometimes from one OS version to the other. And not really, even Xcode. Have you tried downloading the 14.4 simulator image from Xcode and running your App in Xcode 12.5?
